### DISCUSSION TITLE
ring-client-example not capturing Ring Camera events

### DISCUSSION THREAD

**Discussion by cardsfan7189** [dgreif/ring#1403] [Q&A] [answered]
Finally a question about the api and not typescript.  Been running the [ring-client-example](https://github.com/dgreif/ring-client-example) and with debug I can see that the camera details are being captured and code is reaching the camera onNewNotification.subscribe property/method. However when I trigger a motion event or ring the doorbell, nothing happens while the script is running.  This is the last display: "Listening for motion and doorbell presses on your cameras." 

Looked through the issues, discussions and wiki, and the only possible problems I could find were these

- something else is capturing the event before the api does.  Do I need to turn off my Ring app on the phone (Android) ?
- e2e encryption can cause problems.  Checked and this is not enabled

Another question, and maybe this was answered by @tsightler in another discussion: is this api client interacting with the ring camera itself, or is it communicating with events on the Ring-hosted cloud?

Sincerely,

A Cheapskate Trying to Avoid $40 a Year to Ring

**Comment by cardsfan7189** [comment]
And just now saw @tsightler reply to [another discussion](https://github.com/dgreif/ring/discussions/1402#discussioncomment-8509505) about this.  Will look into the token management.

**Reply by tsightler** [reply, marked as answer]
Summary version:

1) Use ring-auth-cli to generate token
2) Use token to connect to API but don't save this initial updated token which now includes Firebase device registration information
3) Push notifications will never work until you delete device from Ring Control Center and regenerate a new token

> Another question, and maybe this was answered by @tsightler in another discussion: is this api client interacting with the ring camera itself, or is it communicating with events on the Ring-hosted cloud?

There's no local communications, Ring is a cloud-only device, the only way to Ring devices is via their cloud API.

> A Cheapskate Trying to Avoid $40 a Year to Ring

Just some friendly advice, sell your Ring camera and buy something else.

**Reply by cardsfan7189** [reply]
Sounds like some good friendly advice 👍 Hopefully my last question, since you've given me plenty to work with:  would I still be able to access camera data and events from their cloud api without the basic subscription?

**Reply by tsightler** [reply]
A subscription is not required, you can do anything that the Ring app allows you to do without a subscription, which is effectively basic notifications, no event history or recordings.  You can get an event, and, if you have a high powered camera (basically a "Pro" doorbell, or other cameras running on AC/mains power), you can take a snapshot or start a livestream.  Battery/low-power cameras can't even do the snapshot because they can't snapshot while recording, and a motion even always starts a recording, however, you can start a live stream and record to a file.

There are already plenty of examples out there of people doing this with ring-mqtt or ring-client-api directly, but it's a significantly sub-par experience vs having a subscription.

**Comment by cardsfan7189** [comment]
> Summary version:
> 
> 1. Use ring-auth-cli to generate token
> 2. Use token to connect to API but don't save this initial updated token which now includes Firebase device registration information
> 3. Push notifications will never work until you delete device from Ring Control Center and regenerate a new token


And finally...item 3 is what did the trick.  It's working now.  [example.ts](https://github.com/dgreif/ring-client-example) seems to be already doing item 2.

**Reply by tsightler** [reply]
Yes, the example project does handle this, however, the point is if the refresh token had ever been used even one time to connect to the API without this, then it will never work and you have to perform step 3 and start over.  Note that, after that, you actually can re-use the token without saving it, but that very first time it's absolutely critical to get the new token because it contains the persistent Firebase registration data and, if you miss it, it's gone forever.

**Comment by danielvand** [comment]
Hi,
I'm getting a little stuck here. based on the example I successfully created a script to turn some lights when somebody rings the doorbell, super convenient. I had to restart the script and I don't get notifications anymore. I had this before, and by remove all devices from Ring I could solve it (step 3). Somehow, not this time. :(

I tried the steps multiple times (restart device, step 1 and 3, script immediately takes care of step 2), even tried to login with my wife's account, I could connect, but don't get any notifications, from any device.

Also I got a second device (raspberry pi) and installed everything from scratch, same result, no notifications. I do see the 2nd device in the ring control center

Am I missing anything here?

**Reply by tsightler** [reply]
Are you using the latest version of the ring-client-api?  If you are using something prior to version 13, the push notification format changed due to the retirement of legacy FCM APIs, so only the latest versions can work.  Also, if you were accessing properties of the push notifications directly, you would need to modify them as the JSON format is now completely different due to this change.

Also, you must always, always, listen for token updates and save new tokens to be used the next time as covered in the Wiki.

Otherwise, you need to make sure you have a connection to mtalk.google.com on port 5228 and that it is resolving the correct server for your region (if you are using DNS adblockers, secure DNS, DNS-over-HTTPS or other things that would cause the geographical load-balancing to fail.

**Reply by danielvand** [reply]
Hi, first of all, thanks for your answer. 
On my 2nd device (to start from scratch) I just cloned https://github.com/dgreif/ring-client-example.git, but I fear that uses the older version of the ring-client-api. Let me try to find out how to clone the latest ring-client-api and get it to work (I'm not so familiar with ts) :)

**Reply by danielvand** [reply]
In the "package.json" and "package-lock.json" of the ring-client-example I updated all ring versions to 13.2.1, unfortunately with the same result. The example doesn;'t use the json directly but the attribute from the ding event.

Token updates are done by the example script and mtalk.google.com on port 5228 is accessible (I've a simple direct internet connection anyways).

**Reply by tsightler** [reply]
Did you start with an entirely new token after updating and have you removed all legacy entries from Ring Control Center?

In the end, the overall operation is simple, ring-client-api registers with FCM, gets a push token, registers with the Ring API as a device, sends Ring this push token, Ring forwards messages via FCM using the token, Ring client API receives the push message over the socket.  The push messages either arrives, or does not.  If messages do not arrive, then you must troubleshoot the token registration process and message receivers.  You can try enabling debug for the API config and see if you get more information back.

Alternatively, many users seem not to realize that modifying notification settings in the Ring app also impacts ring-client-api, so you should make sure you have not disabled any global notifications setting there, things like smart alerts or any other settings which suppress alerts.

**Reply by danielvand** [reply]
Hi,
How do I start with a new token? Because the auth process gets one from the ring api right?
So what I did on a new fresh raspberry was het following steps:

1. Removed all entries from ring control center
2. git clone https://github.com/dgreif/ring-client-example.git
3. sed -i 's/11.7.1/13.2.1/g' package.json 
4. sed -i 's/11.7.1/13.2.1/g' package-lock.json
5. npm i
6. npm run auth
7. Put the created refreshtoken in ".env"
8. npm start

Result:
1. Found all locations and devices
2. got an updated token, which was written to .env
3. Listening for motion and doorbell presses. 
4. Nothing happens when the doorbell is pressed or their is movement. (Ring app and alexa do get notifications)

Any tips what I could try to do a different test?

**Reply by tsightler** [reply]
Every tip that could possibly be written is already in [the wiki](https://github.com/dgreif/ring/wiki/Notification-Troubleshooting).  If those tips don't work for you, then the only option you have is to dig into the code and see if you can figure out why it is failing for you even though it works for thousands of users.  That's not to say it could not be a problem with the code, but the overall process is fairly well understood and quite basic.  The entire push notification registration code is in api.ts so you can add some debug there, but, as I can't reproduce the issue, there's not much more I can do to help other than to point out the common issues that have been seen.
